### PR TITLE
refactor some Socket test code into SocketTestHelper, and modify some existing tests to use this

### PR DESCRIPTION
**Pull request by geoffkizer**
Refactor the existing logic in SendReceive.cs for abstracting across sync/APM/Task/etc and generalize it.  

Also, update existing Accept and Connect tests to use this.  Some existing Accept/Connect tests that are specific to SocketAsyncEventArgs are moved to SocketAsyncEventArgsTests.cs.

PR is split into two commits: first has all the changes, second is just file renaming/reorganization.

@stephentoub @davidsh @CIPop @Priya91 @wfurt

### CODE REVIEW

**Comment by geoffkizer** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

**Comment by geoffkizer** [comment]
@dotnet-bot test Linux x64 Release Build

File: src/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    //
+    // MemberDatas that are generally useful
+    //
+
+    public abstract class MemberDatas
+    {
+        public static readonly object[][] Loopbacks = new[]
+    {
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs]: Nit: indentation

Comment (geoffkizer) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs]: Thanks -- any other feedback?

**Comment by geoffkizer** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop Windows x64 Debug Build

(Failures look unrelated, but running again anyway)

File: src/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs]: LGTM

**Comment by geoffkizer** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

**Comment by geoffkizer** [comment]
Linux failures are known (fedora) or unrelated (httpclient)

**Pull request merged**